src/function/gds/strongly_connected_components_kosaraju.cpp

```diff
@@ -0,0 +1,259 @@
+            for (auto i = 0u; i < numNodes; ++i) {
+                componentIDs[i] = NOT_VISITED;
+            }
+        }
+    }
+
+    void setVisited(const offset_t offset) const { componentIDs[offset] = VISITED; }
+
+    bool visited(const offset_t offset) const { return componentIDs[offset] >= VISITED; }
+
+    void setProcessed(const offset_t offset) const { componentIDs[offset] = PROCESSED; }
+
+    bool processed(const offset_t offset) const { return componentIDs[offset] == PROCESSED; }
+
+    bool componentIDSet(const offset_t offset) const { return componentIDs[offset] < NOT_VISITED; }
+
+    void setComponentID(const offset_t offset, const offset_t value) const {
+        componentIDs[offset] = value;
+    }
+
+    offset_t getComponentID(const offset_t offset) const { return componentIDs[offset]; }
+
+    offset_t getNumNodes() const { return numNodes; }
+
+    table_id_t getTableID() const { return tableID; }
+
+private:
+    offset_t* componentIDs = nullptr;
+    ObjectArraysMap<offset_t> componentIDsMap;
+
+    offset_t numNodes;
+    table_id_t tableID;
+};
+
+class SCCCompute {
+public:
+    SCCCompute(Graph* graph, SCCState& sccState) : graph{graph}, sccState{sccState} {}
+
+    void compute() {
+        auto nbrTables = graph->getForwardNbrTableInfos(sccState.getTableID());
+        if (nbrTables.size() != 1) {
+            throw RuntimeException("Kosaraju's SCC only supports operations on one edge table.");
+        }
+        auto nbrInfo = nbrTables[0];
+        auto relEntry = nbrInfo.relEntry;
+        auto scanState = graph->prepareRelScan(relEntry, "");
+        for (auto i = 0u; i < sccState.getNumNodes(); ++i) {
+            if (!sccState.visited(i)) {
+                forwardDFS(i, *scanState);
+            }
+        }
+        KU_ASSERT(toProcess.size() == 0);
+        for (auto it = dfsStack.end() - 1; it >= dfsStack.begin(); --it) {
+            auto node = *it;
+            if (!sccState.componentIDSet(node)) {
+                backwardsDFS(node, node, *scanState);
+            }
+        }
+    }
+
+    void forwardDFS(offset_t node, NbrScanState& scanState) {
+        toProcess.push_back(node);
+
+        while (!toProcess.empty()) {
+            auto nextNode = toProcess.back();
+            if (sccState.visited(nextNode)) {
+                toProcess.pop_back();
+                if (!sccState.processed(nextNode)) {
+                    dfsStack.push_back(nextNode);
+                    sccState.setProcessed(nextNode);
+                }
+                continue;
+            }
+            sccState.setVisited(nextNode);
+            auto nextNodeID = nodeID_t{nextNode, sccState.getTableID()};
+            for (auto chunk : graph->scanFwd(nextNodeID, scanState)) {
+                chunk.forEach([&](auto nbrNodeID, auto) {
+                    if (!sccState.visited(nbrNodeID.offset)) {
+                        toProcess.push_back(nbrNodeID.offset);
```

Comment:
Well here we are still materializing every nodeID from value vector. I thought the design was to use a `ScanState` pool and not materializing them. Did this change?

Reply:
Copying out the neighbors seem to be the straightforward way.
Deferring this optimization until later when it becomes a bottleneck.

Reply:
I don't know if we want to actually do this at all. It's not really an optimization. This is the standard way to implement iterative DFS-based algorithms.